src/parsers/qasm3_parser/Scanner.cpp

```diff
@@ -0,0 +1,578 @@
+  while (ch != delim) {
+    content << ch;
+    nextCh();
+  }
+
+  t.str = content.str();
+
+  expect(delim);
+
+  t.endCol = col;
+  t.endLine = line;
+
+  return t;
+}
+
+Scanner::Scanner(std::istream& in) : is(in) {
+  keywords["OPENQASM"] = Token::Kind::OpenQasm;
+  keywords["include"] = Token::Kind::Include;
+  keywords["defcalgrammar"] = Token::Kind::DefCalGrammar;
+  keywords["def"] = Token::Kind::Def;
+  keywords["cal"] = Token::Kind::Cal;
+  keywords["defcal"] = Token::Kind::DefCal;
+  keywords["gate"] = Token::Kind::Gate;
+  keywords["opaque"] = Token::Kind::Opaque;
+  keywords["extern"] = Token::Kind::Extern;
+  keywords["box"] = Token::Kind::Box;
+  keywords["let"] = Token::Kind::Let;
+  keywords["break"] = Token::Kind::Break;
+  keywords["continue"] = Token::Kind::Continue;
+  keywords["if"] = Token::Kind::If;
+  keywords["else"] = Token::Kind::Else;
+  keywords["end"] = Token::Kind::End;
+  keywords["return"] = Token::Kind::Return;
+  keywords["for"] = Token::Kind::For;
+  keywords["while"] = Token::Kind::While;
+  keywords["in"] = Token::Kind::In;
+  // TOOD: handle pragma properly
+  keywords["pragma"] = Token::Kind::Pragma;
+  keywords["input"] = Token::Kind::Input;
+  keywords["output"] = Token::Kind::Output;
+  keywords["const"] = Token::Kind::Const;
+  keywords["readonly"] = Token::Kind::ReadOnly;
+  keywords["mutable"] = Token::Kind::Mutable;
+  keywords["qreg"] = Token::Kind::Qreg;
+  keywords["qubit"] = Token::Kind::Qubit;
+  keywords["creg"] = Token::Kind::CReg;
+  keywords["bool"] = Token::Kind::Bool;
+  keywords["bit"] = Token::Kind::Bit;
+  keywords["int"] = Token::Kind::Int;
+  keywords["uint"] = Token::Kind::Uint;
+  keywords["float"] = Token::Kind::Float;
+  keywords["angle"] = Token::Kind::Angle;
+  keywords["complex"] = Token::Kind::Complex;
+  keywords["array"] = Token::Kind::Array;
+  keywords["void"] = Token::Kind::Void;
+  keywords["duration"] = Token::Kind::Duration;
+  keywords["stretch"] = Token::Kind::Stretch;
+  keywords["gphase"] = Token::Kind::Gphase;
+  keywords["inv"] = Token::Kind::Inv;
+  keywords["pow"] = Token::Kind::Pow;
+  keywords["ctrl"] = Token::Kind::Ctrl;
+  keywords["negctrl"] = Token::Kind::NegCtrl;
+  keywords["#dim"] = Token::Kind::Dim;
+  keywords["durationof"] = Token::Kind::DurationOf;
+  keywords["delay"] = Token::Kind::Delay;
+  keywords["reset"] = Token::Kind::Reset;
+  keywords["measure"] = Token::Kind::Measure;
+  keywords["barrier"] = Token::Kind::Barrier;
+  keywords["true"] = Token::Kind::True;
+  keywords["false"] = Token::Kind::False;
+  keywords["im"] = Token::Kind::Imag;
+  //        keywords["dt"]         = Token::Kind::TimeUnitDt;
+  //        keywords["ns"]         = Token::Kind::TimeUnitNs;
+  //        keywords["us"]         = Token::Kind::TimeUnitUs;
+  //        keywords["mys"]        = Token::Kind::TimeUnitMys;
+  //        keywords["ms"]         = Token::Kind::TimeUnitMs;
+  //        keywords["s"]          = Token::Kind::TimeUnitS;
```

Comment:
why are these not enabled here?

Reply:
The `s` keyword was a problem as it collided with the `s` gate, that's why I decided to disable all of these for the moment. I will re-enable them and handle this in the parser.